docs/design/uptime-reputation-calculation.md

```diff
@@ -0,0 +1,122 @@
+)
+```
+
+The table will be filled with _SNs_ which have failed at least one uptime check during the ongoing month.
+
+Each row in the table indicates when an _SN_ failed the first uptime check (`when`), while the _SN_ not succeed any future uptime check (`back_online` is `NULL`); when the _SN_ is back online any future uptime check failure will insert a new row to mark a new period of time that the node is offline again.
+
+When the _SN_ is disqualified by any disqualification system and the _SN_ isn't back online, the `disqualified` column of the row will be updated for not receiving more uptime checks because we don't do business with disqualified _SNs_.
+
+### Algorithm
+
+The algorithm is composed of 2 parts:
+
+1. Select which _SNs_ will be checked for uptime, henceforth uptime check _SN_ selection.
+1. Recheck the uptime of the _SNs_ which has failed the initial uptime check (part 1) during an interval period of time, henceforth uptime recheck loop.
+
+#### Uptime Check _SN_ selection
+
+This design doc doesn't change the _SN_ selection process for uptime checks therefore the current implemented mechanism with some minor modifications to classify those nodes which fail the uptime check in order of being recheck by the second part of the algorithm.
+
+Once an _SN_ is selected for an uptime check, the current implemented uptime check mechanism must be modified to achieve the following algorithm:
+
+1. If there is a row in the `failed_uptime_checks` table for the selected _SN_ with `back_online` set to `NULL` do nothing, otherwise follow with the next step.
+1. Check the _SN_ as it's currently done; if it succeeds, then end, otherwise follow with the next step.
+1. Insert a new row in the `failed_uptime_checks` using its node ID and setting the current timestamp to `when`.
+
+
+#### Uptime Recheck Loop
+
+This process should run independently from any current satellite process and it should run in a configurable time interval.
+
+The algorithm for each time interval iteration is the following:
+
+1. Select all the rows from `failed_uptime_checks` table which has `back_online` and `disqualification` columns set to `NULL` sorted by `when` in ascending order.
+1. For each selected _SN_, retrieve the _SN_ address<sup>1</sup> and performs the uptime check.
+    * If it fails
+        1. Calculate its total offline time of the month, accumulating the offline time interval of each row corresponding to this _SN_ with `back_online` not being `NULL` and the current offline interval time calculated with the current timestamp and the `when` value of this row. If the total offline time doesn't exceed the established network limits end, otherwise follow with the next step.
+        2. Update the `disqualified` column of the row with the current timestamp and disqualifies the _SN_<sup>2</sup>
+    * If it succeeds
+        1. Update the `back_online` column of the row with the current timestamp.
```

Comment:
I see 2 edge case. Lets say we allow an offline time of 24 hours before we disqualify a storage node.

1. A storage node can be offline for 5 minutes every day (dynamic IP change). -> It fails the inital check and will be marked as offline. Now the loop will select the storage node after 1 hour and notice that it is not offline. back_online will be set. At the end of the month the storage node gets disqualified. In our database we have 30 entries with 1 hour each. The storage node was offline only 5 minutes each but we didn't catch that. The problem is that we will disqualify storage nodes because the loop takes to long for one interation. We could ignore the first or last uptime check. If the storage node is back online than set back_online not with the current timestamp. Instead use the last when timestamp. In this example that would mean the storage node gets 0 penalty and we would be force to reduce the loop interval in order to catch him.


2. The satellite might be offline for several hours. We need to make sure satellite offline time doesn't count. How about a job that is running once on restart and sets back_online for all nodes. Same trick. If we set it to the last when timestamp we can make sure the storage nodes are getting 0 penalties for our mistakes.

Reply:
Another option is to fill any `back_online` columns necessary when the node connects with the satellite as well. Then, the node coming online could reach out during startup.

Reply:
@littleskunk I thought about the (1) and let's discuss in the meeting how we can detect those cheaters because the document was trying to define that the if the interval between the last failed uptime check and the next one was ignored for not unfairly penalizing.

Let's discuss the scenario (2) too.

Reply:
As commented in the meeting, I will update the document to reflect that  the uptime check that detects the _SN_ is back online, it will request if the _SN_ has contacted the satellite before this last check and after the previous one and if it did, then use such time stamp to fill the `back_online` otherwise use the current timestamp.

I will add a new column (`last_check`) to register the timestamp of the previous check because it will solve the scenario (2) and will allow to querying the database one row a time rather than having to have an open result set for too long.

Also, I will add a column to count the number of checks despite that it could be roughly calculated knowing the used time interval it's less straightforward and we'll need to track if the time interval has been changed during the period of time involved for the calculation.

--- 

__I've got a question when writing this__.

Once we use the last _SN_ contact timestamp with the satellite or the current timestamp, do we need to discard the elapsed time between the succeeded uptime check and the previous failed one?

If we assume that when an _SN_ goes offline, it has always to contact the satellite at the moment that it gets online again then it looks that we can precisely measure the offline time; if the _SN_ didn't do so we could consider that it's cheating.

Reply:
The columns are added.
I also found that we needed a new table to know the current uptime check period in case the satellite restarts.